Add a linter for rack.early_hints

casperisfine opened the pull request:
Fix: https://github.com/rack/rack/issues/1692

This is kind of a first stab, I'm not familiar enough with `Rack::Lint` to understand exactly what's needed. Please let me know if there are specifications I missed, or if the main description should be improved.

cc @jeremyevans

jeremyevans commented on lib/rack/lint.rb:
```diff
@@ -655,6 +657,29 @@ def check_hijack_response(headers, env)
+        ##
+        ## If rack.early_hints is present it must respond to #call.
+        assert("rack.early_hints must respond to call") { env[RACK_EARLY_HINTS].respond_to?(:call) }
+        original_callback = env[RACK_EARLY_HINTS]
```
This local variable is not used.  It looks like other than checking `env['rack.early_hints']` responds to call, the value is ignored.  Shouldn't `original_callback` be called with the headers inside the lambda below?  Otherwise `Rack::Lint` breaks applications using early hints?

jeremyevans commented on lib/rack/lint.rb:
```diff
@@ -655,6 +657,29 @@ def check_hijack_response(headers, env)
+          ## The headers must respond to #each_pair
+          assert("rack.early_hints headers must respond to each_pair") { headers.respond_to?(:each_pair) }
+          headers.each_pair do |key, value|
+            ## The headers keys and values must respond to #to_s
```
Period at the end of sentences in Lint comments, since they are combined in SPEC.  Otherwise I think RDoc makes them appear as a single sentence.

It would also be better to be more descriptive. `The rack.early_hints argument must respond to #each_pair and yield a key string and a value string to the provided block for each header.`  This would change the specification slightly (`is_a?(String)` instead of `respond_to?(:to_s)`).

ioquatix commented:
I'm okay with this, but I think we should defer to the response headers rather than re-specifying it to avoid duplicating the specifications.

jeremyevans commented:
With the changes to SPEC since this was originally submitted, I think this will require significant changes before merging.  `rack.early_hints` should be called with an argument that would be valid as response headers.  The Lint implementation can call `check_headers` with the value.

@casperisfine Unless you want to work on this, I can see if I can update it tomorrow.

casperisfine commented:
I don't absolutely want to work on it, feel free to take over if you wish, but otherwise I can probably do that in a day or two.

casperisfine commented:
Sorry for the delay, I rebased and replaced the content of `check_early_hints` by `check_headers`.

It's been a while, so I hope the result is decent. Again feel free to directly take over if you'd rather quickly improve it rather than go through review rounds.

casperisfine commented:
Ah that great!
